### DISCUSSION TITLE
Packet duplication (mirror)

### DISCUSSION THREAD

**Discussion by volga629-1** [FoobarOy/foomuuri#35] [General]
Hello Everyone, 
I am trying to mirror  inbound traffic on same machine just different interface.
I can't find good way to achieve  with foomuuri. 

NFtable rule 

```
nft add rule mangle prerouting 'dup to 172.26.200.1 device ips-loop0'

```

**Comment by kimheino** [comment]
This can't be implemented to Foomuuri because nftables only supports "dup to" in ip/ip6 table, not in inet table. Foomuuri generates inet table.

There is still a solution. With nftables you can use multible tables in parallel. Run following shell script to create another table to duplicate packets:

```
#!/usr/sbin/nft -f

table ip duplicator
delete table ip duplicator

table ip duplicator {
	chain dup_to {
		type filter hook prerouting priority mangle
		dup to 172.26.200.1 device ips-loop0
	}
}
```

Foomuuri doesn't know anything about this table and doesn't touch it. You could even use Foomuuri's post_start hook to run it:

```
hook {
   post_start /path/to/my/duplicator
}
```

**Comment by volga629-1** [comment]
Thank you for reply, I was going same route, only my thought was to add duplicator table with NF_IP_PRI_RAW (-300) to RAW chain. 


```
add table ip duplicator
add chain ip duplicator dup_to { type filter hook prerouting priority -300 mangle; }
```

**Reply by volga629-1** [reply]
I tested packet duplication, but with tcpdump I don't see any packets on mirrored device. 

Rules like this, I see match on counter, but no packets on ips-loop0

```
table ip mangle {
        chain prerouting {
                type filter hook prerouting priority mangle; policy accept;
                iifname "ens18" udp dport 5060-5088 counter packets 200 bytes 95060 dup to 172.26.200.1 device "ips-loop0" notrack
        }

        chain input {
                type filter hook input priority mangle; policy accept;
                iifname "ips-loop0" udp dport 5060-5088 ip daddr set 172.26.200.1 notrack
        }
}
```

**Reply by kimheino** [reply]
Does interface ips-loop0 have route to 172.26.200.1?

**Reply by volga629-1** [reply]
Ip 172.26.200.1  is assigned on ips-loop0

**Reply by kimheino** [reply]
Nftables documentation is not that clear, but I assume that with "device ips-loop0" it must be routed. You can try with IP only, without "device XXX".

https://wiki.nftables.org/wiki-nftables/index.php/Duplicating_packets
